### ISSUE TITLE
[feature] rel-me in profile fields

### ISSUE THREAD

**Issue by jeherve** [superseriousbusiness/gotosocial#2216]
### Is your feature request related to a problem ?

Some services rely on the `rel` attributes on profile pages to verify the identity and link identities together.

> Using rel=me on a hyperlink indicates that its destination represents the same person or entity as the current page
-- https://indieweb.org/rel-me

GoToSocial profile fields do not include that attribute, so it can be difficult to verify your Fediverse identity with some services.

### Describe the solution you'd like.

It would be nice if we could add that `me` attribute to all the links in the profile fields.

### Describe alternatives you've considered.

NONE

**Comment by tsmethurst** [comment]
Thanks! Surprised we didn't have an issue for this open already, seems like a handy feature to have!

**Comment by daenney** [comment]
> It would be nice if we could add that `me` attribute to all the links in the profile fields.

I think having the `rel="me"` on links is a nice enhancement, but I don't think we should default to adding that to any/all link on the profile. There's no guarantee that a link someone adds to their profile field points to a website that is "them". Someone can easily add a link to a collective, open source project or something else entirely that they participate in, but isn't "them".

I realise that Mastodon, Pixelfed and others just smash a `rel="me"` on all those links, but that doesn't feel correct, semantically or technically. We should have some way for the user to say "hey, this link points to me on another platform" and for those cases add the `me` to the `rel` on a link. Unfortunately because of Masto's implementation, there's no way for you to do that from a client app, that toggle would have to happen in our web UI.

**Comment by benjaminbhollon** [comment]
> There's no guarantee that a link someone adds to their profile field points to a website that is "them".

The idea to a verify a link is that *both* ends would have an `[rel=me]` link to each other.

Given that, perhaps the path forward would be to (1) do Mastodon-style verification of links to see if the other end has an `[rel=me]` link, then (2) if that link exists on the other end, add `[rel=me]` to the link on the GoToSocial profile page.

That feels like the best balance to me.

**Comment by benjaminbhollon** [comment]
Would be related to #2352 if we took that approach.

**Comment by daenney** [comment]
> > There's no guarantee that a link someone adds to their profile field points to a website that is "them".
> 
> The idea to a verify a link is that _both_ ends would have an `[rel=me]` link to each other.
> 
> Given that, perhaps the path forward would be to (1) do Mastodon-style verification of links to see if the other end has an `[rel=me]` link, then (2) if that link exists on the other end, add `[rel=me]` to the link on the GoToSocial profile page.
> 
> That feels like the best balance to me.

Yup, that seems like it would work fine. We start by retrieving the link, if the target has a link back to us with `[rel=me]` on it then we're good and can add it on our side too. We have to store the result of a link validation and revalidate them from time to time anyway, so at that point we'd have all the data to correctly decide which link gets the attribute.

**Comment by rochacbruno** [comment]
+1 for this issue

I just checked https://indiewebify.me/validate-rel-me/?url=https%3A%2F%2Frochacbruno.com and cannot yet link my GTS with my blog.

**Comment by daenney** [comment]
Please don't +1 post on issues. It creates notification noise. There are emoji reactions on the opening issue, and you can hit the subscribe button in the right sidebar to be notified of updates.

**Comment by ellotheth** [comment]
> Given that, perhaps the path forward would be to (1) do Mastodon-style verification of links to see if the other end has an [rel=me] link, then (2) if that link exists on the other end, add [rel=me] to the link on the GoToSocial profile page.

> We start by retrieving the link, if the target has a link back to us with [rel=me] on it then we're good and can add it on our side too.

A design like this makes the availability and correctness of the target a pretty big dependency:

- I add a profile link to my website, which isn't live yet.
- I add a profile link to my website, which doesn't have a corresponding `rel="me"` to my GTS profile yet.
- I add a profile link to my website, which has a typo in the `rel="me"` statement.
- I add a profile link to my website, which is running on a host that's having an outage.
- I add a profile link to my website, but I messed up the firewall configuration on the host.
- I add a profile link to a third-party profile page that doesn't support `rel="me"`, but I still want the metadata in my GTS profile because I'm a huge nerd.

In all those scenarios, verification fails, and I don't have a clear way to retry or fix it. Is there a button? Do I clear it, then put it back when I'm reasonably sure the target is up and functional? What about the case where there will never be a matching link from the target, but I still want the tag?

As a (new 👋🏻 hi how's it goin, love the project, y'all are super cute) user, I'd rather be able to control that `rel="me"` tag myself than hope a magical system maybe sets it up for me. The current profile link configuration gives us two text fields; could there be a checkbox for each one with some kind of _hey this is me_ label?

![image](https://github.com/user-attachments/assets/d48e93b5-7deb-4484-abc1-125a110bea71)

Maybe add a lil tooltip or help text with a short microformat explainer and a link?

On the other hand,

> There's no guarantee that a link someone adds to their profile field points to a website that is "them".

This is true, but I don't think it matters, and that's probably why other profile pages toss it around with such abandon. The `rel="me"` tag does nothing on its own, so if somebody puts an unrelated site in their profile, the worst-case scenario is "whoops, a robot or nerd saw some incorrect metadata". Verifying that the GTS profile and its target are connected makes sense if you're showing some sort of "verified" indicator, but I don't think it's a good choice to control the presence of the tag itself.

**Comment by tsmethurst** [comment]
I think we'll probably get more clarity on this by implementing https://github.com/superseriousbusiness/gotosocial/issues/2352 *first*, and then coming back to this. At that point, we can decide whether to also add `rel=me` in GoToSocial profile HTML as well, as a way of "completing the loop" so to speak.

**Comment by benjaminbhollon** [comment]
> could there be a checkbox for each one

I'm not sure how clear a flow this would be, considering that a lot of these metadata fields aren't links, and the checkbox wouldn't make sense for them.

That said to put my two cents towards the "re-checking" verification (and by extension whether to have `rel="me"`, especially since GtS tends to be single-user instances, it could make sense to have a button to manually trigger link verification; I remember being annoyed back on Mastodon trying to get it to trigger by removing and re-adding a link. Perhaps with a cooldown to prevent abuse in the cases like my own instance that do have many users.

And yes, I agree, the other issue is probably the one to tackle first; this feels relevant to both, though.

**Comment by ellotheth** [comment]
@benjaminbhollon 

> I'm not sure how clear a flow this would be, considering that a lot of these metadata fields aren't links, and the checkbox wouldn't make sense for them.

Yeah big agree 👍🏻  It's also a hard thing to communicate in a concise way. _Do you want to be able to verify that you own this profile from other websites that also use this microformat?_ is kind of a mouthful. There are no technical downsides to asking for `rel="me"` on a profile entry that's not a link, but it's confusing UX regardless.

> the other issue is probably the one to tackle first

I know you and @tsmethurst and @daenney are all aligned on this so I'll only repeat myself once and then leave you all alone: Globally adding `rel="me"` to links in the profile fields has an immediately positive impact for end users ("hey look, the GTS instance I linked from my Mastodon profile is verified on Mastodon now!"), and no negative impacts on anyone. I'd be happy to stand up a PR with a change to that effect, as an incremental improvement while #2352 is in progress! It might be a great entry into the codebase 😁

**Comment by shombando** [comment]
Interesting discussion here. I was in the same boat here trying to verify a link on my Mastodon profile directing folks to my GoToSocial profile. I look forward to one of these potential implementations.

For now, I'm literally punching in the bare HTML in my profile description and that seems to do the trick on the Mastodon end. In case anyone is else is looking for a bodge:
`Migrating from <a href="https://fosstodon.org/@shom" rel="me">@shom</a>`

**Comment by skyfaller** [comment]
@shombando Thanks for the workaround, I just did this and it worked!

One detail I notice is that this only seems to work in the "Bio" section, not the "Profile fields" where people post their links to get verified on Mastodon. I tried and it stripped the `rel=me`. (Instead it turns it into an `h-card`?)

This also means that if you put the same link in both sections of your profile, one won't have `rel=me`, and this may prevent Mastodon from verifying you.

**Comment by Fastidious** [comment]
For what’s worth, Mastodon client shows GtS profiles links as “verified”, if they have been configured well. 

Using @shombando old instance, for example: https://fosstodon.org/@david@weaknotes.com
